gtsfm/retriever/retriever_base.py

```diff
@@ -0,0 +1,27 @@
+"""Base class for the Retriever, which provides a list of potential image pairs.
+
+Authors: John Lambert
+"""
+
+import abc
+from typing import List, Tuple
+
+from gtsfm.loader.loader_base import LoaderBase
+
+
+class RetrieverBase:
+    def __init__(self) -> None:
+        """ """
+        pass
+
+    @abc.abstractmethod
+    def run(loader: LoaderBase, num_images: int, num_matched: int = 2) -> List[Tuple[int, int]]:
+        """
+        Args:
+            loader: image loader.
+            num_images: total number of images for exhaustive global descriptor matching.
+            num_matched: number of K potential matches to provide per query. These are the top "K" matches per query.
+
+        Return:
+            pair_indices: (i1,i2) image pairs.
+        """
```

Comment:
Should there be a `create_computation_graph()` method to allow for delayed computation like the other frontend modules? Otherwise, I believe it will create an explicit bottleneck in the pipeline.

Reply:
Hi @travisdriver, thanks for taking a look. I believe we would only be able to return a single `Delayed` object, which would not be iterable. Since the retrieval list is the result of a single algorithm call, not separate calls.

So I don't think Dask will work here, at least in our current paradigm.

Reply:
You can address this by returning a dictionary where the pair indices are the keys and the values are the delayed calls to the retriever. This is what's done in #393.

Reply:
Thanks for the reference Travis. In this scenario, if we aren't validating many potential pairs (each validation as a Delayed object), I think we will have an issue. Because we will be returning a single Delayed object (result of 1 algo), and we can't iterate over that. Let me know if I'm misunderstanding, though.

Reply:
You would return `Dict[Tuple[int, int], Delayed]`, which you can iterate over. You could then pass this delayed object to the TwoViewEstimator to be computed when Dask seems fit.

Reply:
Correct me if I'm wrong, but if we generate a NxN similarity matrix, and then select the top entries in each row, we would not be able to wrap the selection for each (i1,i2) pair in Dask's Delayed. This is because generating the entire list is one big Delayed object, and we generate it in a single shot from a single matrix.

I played around with this a decent amount of the retrieval impl, and kept running into this limitation.

Reply:
Yeah there may be situations where the bottleneck is unavoidable. I was trying to avoid explicitly enforcing the bottleneck for all retrievers, but since it's unavoidable for the NetVLAD retriever we can revisit it later if we have to.